File: src/main/java/org/cubexmc/model/RedeemRequirements.java
package org.cubexmc.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Optional preconditions for redeeming a gem.
 */
public class RedeemRequirements {
    public static final RedeemRequirements NONE = new RedeemRequirements(
            Collections.emptyList(),
            true,
            null);

    private final List<RedeemRecipe> recipes;
    private final boolean allowRedeemAll;
    private final String failureMessage;

    public RedeemRequirements(List<RedeemRecipe> recipes, boolean allowRedeemAll, String failureMessage) {
        this.recipes = immutableRecipes(recipes);
        this.allowRedeemAll = allowRedeemAll;
        this.failureMessage = failureMessage;
    }

    /**
     * Legacy constructor kept so older tests and internal callers can migrate gradually.
     */
    public RedeemRequirements(List<String> requiresHeld,
            List<String> requiresRedeemed,
            List<String> consumes,
            List<String> requiresAny,
            int requiresCount,
            List<String> requiresCountFrom,
            boolean allowRedeemAll,
            String failureMessage) {
        RedeemRecipe recipe = RedeemRecipe.legacy(requiresHeld, requiresRedeemed, consumes, requiresAny, requiresCount,
                requiresCountFrom);
        this.recipes = recipe.hasRequirements()
                ? Collections.unmodifiableList(Collections.singletonList(recipe))
                : Collections.emptyList();
        this.allowRedeemAll = allowRedeemAll;
        this.failureMessage = failureMessage;
    }

    private static List<RedeemRecipe> immutableRecipes(List<RedeemRecipe> values) {
        if (values == null || values.isEmpty()) {
            return Collections.emptyList();
        }
        List<RedeemRecipe> copy = new ArrayList<>();
        for (RedeemRecipe recipe : values) {
            if (recipe != null && recipe.hasRequirements()) {
                copy.add(recipe);
            }
        }
        return copy.isEmpty() ? Collections.emptyList() : Collections.unmodifiableList(copy);
    }

    public boolean hasRequirements() {
        return !recipes.isEmpty();
    }

    public List<RedeemRecipe> getRecipes() {
        return recipes;
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getRequiresHeld()}.
     */
    @Deprecated
    public List<String> getRequiresHeld() {
        return flattenIngredientKeys(RedeemRecipe::getRequiresHeld);
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getRequiresRedeemed()}.
     */
    @Deprecated
    public List<String> getRequiresRedeemed() {
        return flattenIngredientKeys(RedeemRecipe::getRequiresRedeemed);
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getConsumes()}.
     */
    @Deprecated
    public List<String> getConsumes() {
        return flattenIngredientKeys(RedeemRecipe::getConsumes);
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getRequiresAny()}.
     */
    @Deprecated
    public List<String> getRequiresAny() {
        if (recipes.isEmpty()) {
            return Collections.emptyList();
        }
        return recipes.get(0).getRequiresAny();
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getRequiresCount()}.
     */
    @Deprecated
    public int getRequiresCount() {
        return recipes.isEmpty() ? 0 : recipes.get(0).getRequiresCount();
    }

    /**
     * @deprecated Use {@link #getRecipes()} and {@link RedeemRecipe#getRequiresCountFrom()}.
     */
    @Deprecated
    public List<String> getRequiresCountFrom() {
        return recipes.isEmpty() ? Collections.emptyList() : recipes.get(0).getRequiresCountFrom();
    }

    public boolean isAllowRedeemAll() {
        return allowRedeemAll;
    }

    public String getFailureMessage() {
        return failureMessage;
    }

    private List<String> flattenIngredientKeys(java.util.function.Function<RedeemRecipe, List<RedeemIngredient>> getter) {
        if (recipes.isEmpty()) {
            return Collections.emptyList();
        }
        List<String> result = new ArrayList<>();
        for (RedeemRecipe recipe : recipes) {
            for (RedeemIngredient ingredient : getter.apply(recipe)) {
                for (int i = 0; i < ingredient.getAmount(); i++) {
                    result.add(ingredient.getGemKey());
                }
            }
        }
        return Collections.unmodifiableList(result);
    }
}

File: src/main/java/org/cubexmc/model/RedeemRequirements.kt
package org.cubexmc.model

import java.util.Collections

/**
 * Optional preconditions for redeeming a gem.
 */
class RedeemRequirements {
    val recipes: List<RedeemRecipe>
    val isAllowRedeemAll: Boolean
    val failureMessage: String?

    constructor(recipes: List<RedeemRecipe>?, allowRedeemAll: Boolean, failureMessage: String?) {
        this.recipes = immutableRecipes(recipes)
        this.isAllowRedeemAll = allowRedeemAll
        this.failureMessage = failureMessage
    }

    constructor(
        requiresHeld: List<String>?,
        requiresRedeemed: List<String>?,
        consumes: List<String>?,
        requiresAny: List<String>?,
        requiresCount: Int,
        requiresCountFrom: List<String>?,
        allowRedeemAll: Boolean,
        failureMessage: String?,
    ) {
        val recipe = RedeemRecipe.legacy(requiresHeld, requiresRedeemed, consumes, requiresAny, requiresCount, requiresCountFrom)
        this.recipes = if (recipe.hasRequirements()) {
            Collections.unmodifiableList(Collections.singletonList(recipe))
        } else {
            emptyList()
        }
        this.isAllowRedeemAll = allowRedeemAll
        this.failureMessage = failureMessage
    }

    fun hasRequirements(): Boolean = recipes.isNotEmpty()

    @Deprecated("Use recipes and RedeemRecipe.requiresHeld.")
    fun getRequiresHeld(): List<String> = flattenIngredientKeys { it.requiresHeld }

    @Deprecated("Use recipes and RedeemRecipe.requiresRedeemed.")
    fun getRequiresRedeemed(): List<String> = flattenIngredientKeys { it.requiresRedeemed }

    @Deprecated("Use recipes and RedeemRecipe.consumes.")
    fun getConsumes(): List<String> = flattenIngredientKeys { it.consumes }

    @Deprecated("Use recipes and RedeemRecipe.requiresAny.")
    fun getRequiresAny(): List<String> = if (recipes.isEmpty()) emptyList() else recipes[0].requiresAny

    @Deprecated("Use recipes and RedeemRecipe.requiresCount.")
    fun getRequiresCount(): Int = if (recipes.isEmpty()) 0 else recipes[0].requiresCount

    @Deprecated("Use recipes and RedeemRecipe.requiresCountFrom.")
    fun getRequiresCountFrom(): List<String> = if (recipes.isEmpty()) emptyList() else recipes[0].requiresCountFrom

    private fun flattenIngredientKeys(getter: (RedeemRecipe) -> List<RedeemIngredient>): List<String> {
        if (recipes.isEmpty()) return emptyList()
        val result = ArrayList<String>()
        for (recipe in recipes) {
            for (ingredient in getter(recipe)) {
                for (i in 0 until ingredient.amount) {
                    result.add(ingredient.gemKey)
                }
            }
        }
        return Collections.unmodifiableList(result)
    }

    companion object {
        @JvmField
        val NONE: RedeemRequirements = RedeemRequirements(emptyList(), true, null)

        private fun immutableRecipes(values: List<RedeemRecipe>?): List<RedeemRecipe> {
            if (values.isNullOrEmpty()) return emptyList()
            val copy = ArrayList<RedeemRecipe>()
            for (recipe in values) {
                if (recipe.hasRequirements()) copy.add(recipe)
            }
            return if (copy.isEmpty()) emptyList() else Collections.unmodifiableList(copy)
        }
    }
}
